Remove 'here' and make method name resolution case-sensitive

hubertp opened the pull request:
### Pull Request Description

Modified UppercaseNames to now resolve methods without an explicit `here` to point to the current module.
`here` was also often used instead of `self` which was allowed by the compiler.
Therefore UppercaseNames pass is now GlobalNames and does some extra work -
it translated method calls without an explicit target into proper applications.

### Important Notes

There was a long-standing bug in scopes usage when compiling standalone expressions.
This resulted in AliasAnalysis generating incorrect graphs and manifested itself only in unit tests
and when running `eval`, thus being a bit hard to locate.
See `runExpression` for details.

Additionally, method name resolution is now case-sensitive.

Obsolete passes like UndefinedVariables and ModuleThisToHere were removed. All tests have been adapted.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

hubertp committed:
Support for method references without 'here'

Added another pass which mimics UppercaseNames pass in a sense that it
adds an explicit `self` of the module whenever it encounteres a variable
that is really a method.

Next step is to translate all `here` to the new construct and remove
`here` as a keyword.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Passes.scala:
```diff
@@ -56,6 +56,8 @@ class Passes(
       ExpressionAnnotations,
       AliasAnalysis,
+      MethodReferences,
+      AliasAnalysis,
```
Off-topic: Shouldn't these passes be collected via `ServiceLoader.load`? Then the system would become modular & extensible. Individual passes could be separated into own compilation units. Some of them could be skipped (during tests or even in runtime).

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/MethodReferences.scala:
```diff
@@ -0,0 +1,201 @@
+    )
+    val freshNameSupply = inlineContext.freshNameSupply.getOrElse(
+      throw new CompilerError(
+        "No fresh name supply passed to UppercaseNames resolver."
```
This is no longer `UpercaseNames` resolver.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/MethodReferences.scala:
```diff
@@ -0,0 +1,201 @@
+    definition match {
+      case asc: IR.Type.Ascription => asc
+      case a =>
+        a.mapExpressions(processExpression(_, bindings, freshNameSupply))
```
Off-topic (a bit): As far as I know `mapExpressions` always creates a `copy` - while it should only create a copy if the underlying references changed - e.g. if there is an identity function sent to `mapExpressions`, no copying shall happen at all.

hubertp replied:
That is a valid concern. I think we could short circuit it by checking if the result of the application of the function returns the same object and not copy if that is the case

hubertp committed:
PR feedback + tests

hubertp committed:
nits

hubertp committed:
Simplify, drop Uppercase logic

hubertp committed:
Remove here from instrumentation tests

Minor caveat - had to make sure that location is properly assigned to
synthetically constructed applications so that instrumentation correctly
picks up nodes.
Also removed all (now) unnecessary `this.` usages; looks like they were
used interchangeably with `here`.

hubertp committed:
s/here.//

kustosz reviewed (changes requested):
This looks nice, but it doesn't deal with the full extent of leftover garbage :) As we've discussed at the beginning, this should remove the `UppercaseNames` pass (or integrate it with your new pass, under a common name `GlobalNames` – the `Uppercase` distinction makes no sense anymore. 
Should also remove the `isReferent` field on `IR.Name`, and everything that uses it. And also all cases of case-conversion (there's a bunch of these in `BindingsMap`).
We're changing language semantics here, keeping the old logic is very confusing and will inevitably lead to bugs.

hubertp committed:
Fix invalid references due to case-insensitivity

Previously binding process used lower-case names of symbols,
irrespective of whether a declaration used capitalized name or not.
This allowed for minor inconsitencies that were resolved by the compiler
irrespective of this mismatch.
Since we are going to be trully case-sensitive, they had to be fixed.

hubertp committed:
Make Uppercase pass case-sensitive, adapt stdlib

Adapted stdlib and tests where we apparently (ab-)used that pattern
frequently.

hubertp committed:
Fix a bug in scope cloning

Merging the passes and depending on AliasAnalysis info (or rather its
scope) revealed what appears to be a long-standing, deeply hidden bug in
scope cloning.
Previously, cloning a scope created a new one that was in no way
associated with the ones present in the graph's chain of scopes. That
meant that various names were not found.
While we managed to somehow avoid this problem, UppercaseNames now
heavily relies on this information being correct.
Notice that for example ChangesetBuilderTest has a weird casting cause
by what appears to be a bug in DemandAnalysis.

hubertp committed:
post rebase fixes

hubertp committed:
referant is gone

hubertp committed:
s/UppercaseNames/GlobalNames

hubertp committed:
Remove redundant ModuleThisToHere pass

Also fixed one test which apparently still used (invalid) semantics

hubertp committed:
Remove (now redundant) UndefinedVariables pass

Depreceated by checks done in GlobalNames pass.

hubertp committed:
Goodbye 'here'

'here' is no longer a keyword.
Using None for typePointer in MethodReference when encountering a local
method from the current module forced minor refactoring.

hubertp committed:
Changelog

hubertp committed:
Fix tail calls for new method resolution

The @Tail_Call annotation would be assigned to the function literal
rather than the application, leading to a warning and lack of
optimisation (and SO in benchmarks)

hubertp committed:
linter

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala:
```diff
@@ -2245,13 +2230,14 @@ object IR {
       */
     sealed case class MethodReference(
-      typePointer: IR.Name,
+      typePointer: Option[IR.Name], // will need to get rid of /Here
```
what's this comment

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala:
```diff
@@ -169,7 +173,8 @@ case class BindingsMap(
   private def findExportedCandidatesInImports(
-    name: String
+    name: String,
+    caseSensitive: Boolean
```
What needs this? I can't think of a use case for `caseSensitive=false`?

hubertp replied:
leftover, thanks for noticing

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/AliasAnalysis.scala:
```diff
@@ -554,21 +612,24 @@ case object AliasAnalysis extends IRPass {
 
-    if (isInPatternContext && name.isVariable) {
+    val resolution = bindings.resolveName(name.name)
+    val isVariable = !name.isMethod && !isTpeConstructor(resolution)
```
I'm very uneasy reading this. What does it matter for alias analysis that something resolved to `ResolvedConstructor`? It shouldn't. Thinking about it even deeper – why does `AliasAnalysis` care about global resolution at all? It should just always try to bind the name to a local symbol and if it fails assume it's global. The old version of this was using a syntactic distinction, we need to carefully think about the new one. Unless I'm missing something, but from my knowledge, it makes no logical sense for the condition to be what it is.

hubertp replied:
Yea, this was unnecessarily complex. I was mimicking the old logic and it turns out there was a much easier solution.

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/BindingAnalysis.scala:
```diff
@@ -76,19 +76,21 @@ case object BindingAnalysis extends IRPass {
             else None
-          case IR.Name.Here(_, _, _) => Some(ref.methodName.name)
-          case IR.Name.Literal(n, _, _, _, _, _) =>
+          //case Some(IR.Name.Here(_, _, _)) => Some(ref.methodName.name)
```
```suggestion
```

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/GlobalNames.scala:
```diff
@@ -129,23 +129,38 @@ case object UppercaseNames extends IRPass {
-
+                lit
+                  .updateMetadata(this -->> BindingsMap.Resolution(r))
+                  .copy(isMethod = true)
```
Very strange to modify `isMethod` like this. `isMethod` is a syntactic marker for alias analysis to not try to resolve this name. Shouldn't ever change it.

hubertp replied:
Yeap, leftover from the incremental changes that led to the final solution

kustosz reviewed (changes requested):
In certain points it seems like it doesn't work or it works by accident. Details left inline, approving after these are resolved by proving me wrong or changing the code :P

hubertp committed:
Address PR review

Got rid of the ugly logic in AliasAnalysis that used resolution info.

JaroslavTulach reviewed (approved):
That's enormous refactoring of the Scala code! I cannot claim I understand all the changes, but it looks sane and if it glues together, let's use it!

hubertp committed:
minor

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Encoding.enso:
```diff
@@ -14,7 +14,7 @@ all_character_sets =
 all_encodings : Vector Encoding
 all_encodings =
-    here.all_character_sets . map Encoding
+    all_character_sets . map Encoding
```
```suggestion
    all_character_sets.map Encoding
```

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time.enso:
```diff
@@ -254,7 +254,7 @@ type Time
              example_zone = Time.now.zone
     zone : Zone
-    zone = Zone.zone (self . internal_zoned_date_time . getZone)
+    zone = Zone.Zone (self . internal_zoned_date_time . getZone)
```
```suggestion
    zone = Zone.Zone self.internal_zoned_date_time.getZone
```

jdunkerley commented on test/Tests/src/Runtime/Stack_Traces_Spec.enso:
```diff
@@ -4,17 +4,17 @@ import Standard.Test
 spec = Test.group "Stack traces" <|
     Test.specify "should capture traces correctly" <|
-        modname = Meta.Constructor (Meta.meta here . constructor) . name
+        modname = Meta.Constructor (Meta.meta self . constructor) . name
```
```suggestion
        modname = Meta.Constructor (Meta.meta self.constructor) . name
```

radeusgd replied:
This change seems to change the semantics, I think the original code was purposefully doing `(Meta.meta self).constructor` because the `constructor` is a method on the `Meta` type.

radeusgd replied:
Does this code work correctly? Looks slightly suspicious. Cannot we use `Meta.meta Stack_Traces_Spec . constructor`? With `here` removed, I guess we still need some way to refer to the current module, right?

hubertp replied:
> This change seems to change the semantics, I think the original code was purposefully doing `(Meta.meta self).constructor` because the `constructor` is a method on the `Meta` type.

Yes, that seemed to have been on purpose

hubertp replied:
Changes to spec name. Thanks for the suggestion, @radeusgd!

jdunkerley commented on test/Visualization_Tests/src/Id_Spec.enso:
```diff
@@ -21,7 +21,7 @@ spec = Test.group "Serializable Visualization Identifiers" <|
             lib = Json.from_pairs [["name", p_name]]
             Json.from_pairs [["library", lib], ["name", v_name]]
-        v_1 = Visualization.Id.from_module here "My Vis"
+        v_1 = Visualization.Id.from_module self "My Vis"
```
Not sure on this but if it passes one to revisit later.

radeusgd replied:
```suggestion
        v_1 = Visualization.Id.from_module Id_Spec "My Vis"
```

Cannot we refer to the current module just be its name like so?

hubertp replied:
Thanks @radeusgd, that works

jdunkerley reviewed (approved):
Only looked at .enso code.
Looks really tidy as always. A couple of random possible fixes.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Noise/Generator.enso:
```diff
@@ -30,7 +30,7 @@ type Generator
        it usually depends on the generator and its intended use.
     step : Number -> Interval -> Any
-    step _ _ = Unimplemented "Only intended to demonstrate an interface."
+    step _ _ = unimplemented "Only intended to demonstrate an interface."
```
I wonder what this `unimplemented` reference points to? I'm pretty sure that it would crash at runtime with method not defined.

I think it needs to be
```suggestion
    step _ _ = Errors.unimplemented "Only intended to demonstrate an interface."
```
with added `import Standard.Base.Error.Common as Errors` like in other places, for example `distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine.enso`.

jdunkerley replied:
This will never have worked. Think Radek's suggestion is the right one.

kustosz replied:
Wrong, it probably was working. `Main` does `from Error.Common export all`. This includes the `unimplemented` method, as long as you call it uppercased. Therefore it will work now in a much less surprising way.

radeusgd replied:
Oh, wow. I didn't know `export all` was able to export methods. Nice.

hubertp replied:
Yeap, that was long time coming

radeusgd commented on distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Database.enso:
```diff
@@ -53,5 +53,5 @@ connect url user=Nothing password=Nothing custom_properties=[] =
 open_sqlite_file : File -> Connection ! Sql_Error
 open_sqlite_file file =
     url = "sqlite:" + file.absolute.path
-    here.connect url
+    connect url
```
This method has been removed already. The merge conflicts need to be resolved.

kustosz commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json/Internal.enso:
```diff
@@ -323,6 +323,6 @@ into_helper fmt json = case fmt of
 parse_helper : Text -> Json
 parse_helper json_text =
-    consumer = here.mk_consumer
+    consumer = mk_consumer
```
My name is Referential Transparency and welcome to Jackass :D

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala:
```diff
@@ -144,16 +144,20 @@ case class BindingsMap(
-      .filter(_.name.toLowerCase == name.toLowerCase)
+  private def findMethodCandidates(
+    name: String,
+    caseSensitive: Boolean
```
one more leftover of this arg

kustosz commented on engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/AliasAnalysis.scala:
```diff
@@ -548,27 +561,30 @@ case object AliasAnalysis extends IRPass {
     * @param name the name to analyse
     * @param isInPatternContext whether or not the name is occurring in a
     *                           pattern context
+    * @param isConstructorNameInPatternContext whether or not the name is a constructor name occurring in a
```
Strange formatting, long line

kustosz reviewed (approved):
Right, this is exactly what it should be! Great work :)

hubertp committed:
Apply suggestions from code review

Co-authored-by: James Dunkerley <jdunkerley@users.noreply.github.com>

hubertp committed:
rebase tweaks

hubertp committed:
PR review

hubertp committed:
rebase tweaks

hubertp committed:
fix formatting

hubertp committed:
PR review

hubertp committed:
nit

hubertp commented:
Thanks y'all. Good feedback.
